### PR TITLE
Dynamic post-processing actions

### PR DESCRIPTION
**Pull request by anatskiy**
The PR adds a new feature by executing a custom Python function after a workflow run. In this example, Galaxy sends a simple desktop notification:

screenshot from 2018-02-01 20-41-13

screenshot from 2018-02-01 20-42-45

All credits go to @bgruening, I just adapted the 3-year-old feature to the latest dev.

### CODE REVIEW

**Comment by bgruening** [comment]
@dannon do you remember our discussion about this extension? 3 years ago we were able to control a robot depending on the output of a workflow, sample new data and resubmit a workflow. A feedback loop was with this possible :)



https://github.com/bgruening/hackathonfr

File: config/galaxy.yml.sample [hunk 1]
```diff
@@ -1,21 +1,21 @@
 # Galaxy is configured by default to be usable in a single-user development
```
Comment (martenson) [inline config/galaxy.yml.sample]: this file is [generated](manage-script-name) from [schema](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/config_schema.yml).

**Review by martenson — CHANGES_REQUESTED** [review]
please remove the `static/script` files and rebase, we no longer track these artifacts in git

**Comment by anatskiy** [comment]
@martenson for some reason I'm unable to revert the two remaining static files. The error is:
```
❯ git checkout e591288cadf5af3acb0700f01cbac1188f07abb1 static/scripts/libs/bibtex.js           
error: pathspec 'static/scripts/libs/bibtex.js' did not match any file(s) known to git.
```

**Comment by martenson** [comment]
@anatskiy I will have a look. You also need to generate the sample config (and commit it) using the makefile target I linked.

Thanks for the adjustments!

edit: I think you can remove those static files in another commit, and then squash it using rebase.

**Comment by anatskiy** [comment]
@martenson what command generates the config file?

**Comment by martenson** [comment]
@anatskiy `make config-rebuild` - sorry for including wrong link previously

**Comment by jmchilton** [comment]
Clever with the docstring - I like that - looks good to me! I also like the client side piece is a light touch - this basically doesn't constrain frontend development at all. Thanks!

**Review by martenson — DISMISSED** [review]

**Comment by jmchilton** [comment]
Rerunning the qunit tests - those sometimes fail randomly but it looked like it might be related to the PR.

**Comment by jmchilton** [comment]
@galaxybot test this

File: client/galaxy/scripts/mvc/workflow/workflow-forms.js [hunk 2]
```diff
@@ -194,8 +194,10 @@ function _visit(head, head_list, output_id, options) {
 /** Builds sub section with step actions/annotation */
 function _makeSection(output_id, options) {
     var extensions = [];
+    var dynamicActionOptions = [];
     var input_terminal_names = [];
     var datatypes = options.datatypes;
+    var dynamicActions = options.dynamicActions;
```
Comment (guerler) [inline client/galaxy/scripts/mvc/workflow/workflow-forms.js:200]: This is correct convention but in this file we are using underscores for variable names i.e. "dynamic_actions" and "dynamic_action_options". Can you adjust this for consistency?

File: lib/galaxy/jobs/actions/post.py [hunk 3]
```diff
@@ -7,12 +7,49 @@
...
+    for action_name in config_post_actions:
+        if ":" in action_name:
+            # Should be a submodule of actions (e.g. examples:microscope_control)
+            (module_name, class_name) = action_name.rsplit(":", 1)
+            module_name = 'galaxy.jobs.actions.dynamic.%s' % module_name.strip()
+            module = __import__(module_name, globals(),
+                                fromlist=['temp_module'])
+            class_object = getattr(module, class_name.strip())
+
+            print(class_object, module)
```
Comment (guerler) [inline lib/galaxy/jobs/actions/post.py:28]: Debug output?

File: lib/galaxy/webapps/galaxy/api/dynamic_actions.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        dynamic_actions = get_dynamic_post_processing_actions(
+            trans.app.config.dynamic_post_processing_actions)
+
+        # Only actions with the correct privileges should be shown
+        dynamic_actions_filtered = []
+        for dynamic_action in dynamic_actions:
+            action_class = dynamic_action['cl'](trans=trans)
+            if action_class.check_privileges():
+                dynamic_actions_filtered.append(dynamic_action)
+
```
Comment (guerler) [inline lib/galaxy/webapps/galaxy/api/dynamic_actions.py:22]: Can you move the privilege check to `get_dynamic_post_processing_actions`? The attribute names should also be returned consistently as either `short_desc` or `name` then this api function might just contain: 
```return get_dynamic_post_processing_actions(trans.app.config.dynamic_post_processing_actions)```.

**Comment by bgruening** [comment]
Thanks all for the review. @RJMW any comments from your side?

**Comment by jmchilton** [comment]
I think the qunit tests are failing because of changes in client/galaxy/scripts/mvc/workflow/workflow-forms.js. Sorry to pile on @anatskiy.

**Pull request closed without merging**